### PR TITLE
Cleaner API in OverlapDetector + added tests

### PR DESCRIPTION
**Pull request by akiezun**
### Description

There's often a need to quickly find a Locatable that overlaps a given static set of Locatables (eg, exons overlapping a given read, or variants overlapping a given read etc). I adapted the OverlapDetector APIs to this usecase. Thanks to @tfenne for pointing me to it and providing examples.

(Note: comments below reflect an old state of this PR - when there was a SkipList implementation competing with the OverlapDetector)

### Checklist

- [x] Code compiles correctly
- [x] New tests covering changes and new functionality
- [x] All tests passing

### CODE REVIEW

**Comment by droazen** [comment]
@akiezun Have you confirmed that using this class resolves the Hadoop-BAM performance issue?

**Comment by akiezun** [comment]
ha, good point. No, I didn't - huge pain to do 3-repo update

**Comment by droazen** [comment]
@akiezun Can you find a way to do that first, to be sure that the existing `SkipList` implementation we're putting in htsjdk actually solves the original issue? Should be possible using custom-built jars and `mavenLocal()`

**Comment by akiezun** [comment]
will do, @cmnbroad can you help me out?

**Comment by tfenne** [comment]
Have you guys benchmarked this vs. the existing interval tree based `OverlapDetector` class?  I'm curious if it's significantly faster for some set of use cases, and if so perhaps we should refactor `OverlapDetector` to be more generate (i.e. `Locatable` based vs. `Interval`) and then also use this code?

**Comment by akiezun** [comment]
@tfenne can you help me with a benchmark? Im not sure if I understand the API of `OverlapDetector`.  How do I use `OverlapDetector` for the following case: I have a fixed set (say, ten thousands) of intervals and I want to query another interval for overlapping elements of the set?

**Comment by tfenne** [comment]
@akiezun Sure - I'm happy to help.  The `OverlapDetector` api is a little clunky, but it has a couple of advantages:
1. It allows you to specify "padding" to intervals (this is obviously of fairly limited usage)
2. It's mutable - so you can add and remove intervals to it as you go
3. It allows you to associate arbitrary objects to each interval that should be returned as the result of the overlap detection, instead of the overlapped interval

But basically you do stuff like this (from Picard's TargetMetricsCollector):
```java
// Setup the overlap detector
final List<Interval> uniqueTargets = IntervalList.fromFile(someFile).uniqued().getIntervals();
final OverlapDetector<Interval> targetDetector = new OverlapDetector<Interval>(0,0);
targetDetector.addAll(uniqueTargets, uniqueTargets);

// Query it
final Collection<Interval> targets = targetDetector.getOverlaps(read);
```

Happy to help with a benchmark - but it should be relatively simple give that `Interval` implements `Locatable`.

**Comment by akiezun** [comment]
perfect, i should be all set for a bakeoff.

**Comment by cmnbroad** [comment]
To use local Hadoop-BAM with GATK, bump the Hadoop-BAM version in pom.xml:
<version>7.4.1-SNAPSHOT</version>
to something newer, then run "mvn install". Then add "mavenLocal()" to the build.gradle repo list; change GATK build.gradle to use the new Hadoop-BAM version number, and build.

If you also need to first *compile* Hadoop-BAM with the new local htsjdk because you're calling new htsjdk apis from Hadoop-BAM, then you need to hack/cheat a bit since you can't locally publish htsjdk. Let me know if you need to do that and I can dig up the details of how I've done it but its a little hacky.

File: src/java/htsjdk/samtools/util/LocatableSkipList.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
+package htsjdk.samtools.util;
```
Comment (nh13) [inline src/java/htsjdk/samtools/util/LocatableSkipList.java]: license

**Comment by akiezun** [comment]
OK, results are in, OverlapDetector won in this benchmark (see OverlapBakeoff.java for code).
I used this for intervals: ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/b37/Broad.human.exome.b37.interval_list.gz

Comments and suggestions welcome.  @droazen @jean-philippe-martin @tfenne @tomwhite does this look like a fair benchmark representative of what we need?

Here's a run with a bam file of 2.0GB as a source of reads to lookup.

results are in milliseconds
```
OverlapDetector
setup time:177
query time:1517
readCount:15746314 counts:{0=1.5307449E7, 1=437651.0, 2=1214.0}

LocatableSkipList
setup time:135
query time:3996
readCount:15746314 counts:{0=1.5307449E7, 1=437651.0, 2=1214.0}
```

another run with a smaller file 1.2GB and more overlaps (exome bam of chr10)
```
OverlapDetector
setup time:165
query time:1578
readCount:9895920 counts:{0=3445591.0, 1=6449016.0, 2=1313.0}

LocatableSkipList
setup time:124
query time:3127
readCount:9895920 counts:{0=3445591.0, 1=6449016.0, 2=1313.0}
```

**Comment by tfenne** [comment]
@akiezun I took a quick look at the benchmark code and I think it's totally idiomatic of how we use such classes, and so is likely a good benchmark.

Maybe we should tidy up OverlapDetector to make it easier to work with, apply the unit tests from this PR to it, and call it a day?

**Comment by jean-philippe-martin** [comment]
I'm sure it's fair - had I known about `OverlapDetector` I'd probably have used that instead of implementing my own.

**Comment by akiezun** [comment]
@ftenne that's my plan. PR coming soon

On Wednesday, April 13, 2016, Tim Fennell <notifications@github.com> wrote:

> @akiezun <https://github.com/akiezun> I took a quick look at the
> benchmark code and I think it's totally idiomatic of how we use such
> classes, and so is likely a good benchmark.
>
> Maybe we should tidy up OverlapDetector to make it easier to work with,
> apply the unit tests from this PR to it, and call it a day?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/samtools/htsjdk/pull/567#issuecomment-209649304>
>


-- 
Sent from Gmail Mobile

**Comment by akiezun** [comment]
@tfenne see the last commit. WDYT? It's a bit faster now too. If people are OK with this, then I'll remove the skip list commits and we'll keep the overlapdetector with new tests.

@tomwhite I changed the API to `Locatable` and added a `overlapsAny` method for a quick check if anything overlaps.
```
OverlapDetector
setup time:230
query time:1349
readCount:9895920 counts:{0=3445591.0, 1=6449016.0, 2=1313.0}
```

**Comment by akiezun** [comment]
@tfenne can you review?

**Comment by yfarjoun** [comment]
@akiezun I think that this PR as it stands it a bit difficult to review. Could you clean it up and bring it to the state that you propose merging? from the comments it seems that the reviewer would need to ignore some stuff but look at other stuff...that's not going to work.

**Comment by akiezun** [comment]
no probs

**Comment by akiezun** [comment]
@yfarjoun this is cleaned up and ready now. To whom should we assign the review?

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java [hunk 2]
```diff
@@ -23,21 +23,22 @@
...
+ * <pre>{@code
+ *    OverlapDetector<Locatable> detector = OverlapDetector.create(locatables);
+ *    Set<Locatable> overlaps = detector.getOverlaps(query);
+ *
+ *    boolean anyOverlap = detector.overlapsAny(query); //faster API for checking presence of any overlap
+ * }</pre>
  */
 public class OverlapDetector<T> {
-    private Map<Object, IntervalTree<Set<T>>> cache = new HashMap<Object, IntervalTree<Set<T>>>();
+    private Map<Object, IntervalTree<Set<T>>> cache = new HashMap<>();
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: is there a reason this isn't final?

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java [hunk 3]
```diff
@@ -53,74 +54,140 @@ public OverlapDetector(int lhsBuffer, int rhsBuffer) {
...
+     *      ! getOverlaps(locatable).isEmpty()
+     * }</pre>
+     */
+    public boolean overlapsAny(final Locatable locatable) {
+        if (locatable == null){
+            throw new IllegalArgumentException("null locatable");
+        }
+        final String seqId = locatable.getContig();
+        final IntervalTree<Set<T>> tree = this.cache.get(seqId);
+        if (tree == null){
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: space

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java [hunk 4]
```diff
@@ -53,74 +54,140 @@ public OverlapDetector(int lhsBuffer, int rhsBuffer) {
...
+    /**
+     * Returns true iff the given locatable overlaps any locatable in this detector.
+     *
+     * This is a performance shortcut API functionally equivalent to:
+     * <pre>{@code
+     *      ! getOverlaps(locatable).isEmpty()
+     * }</pre>
+     */
+    public boolean overlapsAny(final Locatable locatable) {
+        if (locatable == null){
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: space

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java [hunk 5]
```diff
@@ -53,74 +54,140 @@ public OverlapDetector(int lhsBuffer, int rhsBuffer) {
...
-        int end = rhs.getEnd() - this.rhsBuffer;
-
-        if (tree != null && start <= end)
-        {
-            Iterator<IntervalTree.Node<Set<T>>> it = tree.overlappers(start, end);
-            while (it.hasNext())
-            {
-                IntervalTree.Node<Set<T>> node = it.next();
-                matches.addAll(node.getValue());
+        if (start > end){
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: space

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java [hunk 6]
```diff
@@ -53,74 +54,140 @@ public OverlapDetector(int lhsBuffer, int rhsBuffer) {
...
+            return false;
+        }
+
+        final Iterator<IntervalTree.Node<Set<T>>> it = tree.overlappers(start, end);
+        while (it.hasNext()) {
+            final IntervalTree.Node<Set<T>> node = it.next();
+            if (!node.getValue().isEmpty()){
```
Comment (yfarjoun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: space

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+    }
+
+    @Test(dataProvider = "intervalsMultipleContigs")
+    public void testOverlap(List<Locatable> input, Locatable query, Collection<Locatable> expected) throws Exception {
+        final OverlapDetector<Locatable> targetDetector = new OverlapDetector<>(0, 0);
+        targetDetector.addAll(input, input);
+
+        Collection<Locatable> actual = targetDetector.getOverlaps(query);
+        Assert.assertEquals(
+                actual,
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: no need for multiple lines here.

Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: no need for variable either (or make final)

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import java.util.*;
+
+public class OverlapDetectorTest {
+
+    @DataProvider(name="intervalsMultipleContigs")
+    public Object[][] intervalsMultipleContigs(){
+        List<Locatable> input = Arrays.asList(
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: finals all around.

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+        Collection<Locatable> actual = targetDetector.getOverlaps(query);
+        Assert.assertEquals(
+                actual,
+                expected
+        );
+    }
+
+    @DataProvider(name="intervalsSameContig")
+    public Object[][] intervalsSameContig(){
+        List<Locatable> input = Arrays.asList(
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: finals

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+                new Interval("1",10,60),
+                // equal to query interval
+                new Interval("1",30,50),
+                // covered by query interval
+                new Interval("1",40,42),
+                // ends after query interval
+                new Interval("1",45,60)
+        );
+        // returns input single SimpleInterval, query range, expected SimpleInterval
+        return new Object[][]{
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: space

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+                // single-point boundary cases
+                new Object[]{input, new Interval("1", 10, 10), input},
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: spaces.

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+                new Object[]{input, new Interval("1", 10, 100), input},
+                // multiple intervals in the same place (potential edge case for indexing)
+                new Object[]{manyOverlapping, new Interval("1", 20, 20), manyOverlapping},
+                // input with multiple intervals
+                new Object[]{mixInput, new Interval("1",30,50), mixExpected},
+                // input with multiple intervals , non overlapping query
+                new Object[]{mixInput, new Interval("1",300,500), empty},
+        };
+    }
+
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: remove extra newline

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+
+    @Test(dataProvider = "intervalsSameContig")
+    public void testOverlap(List<Locatable> input, Interval query, List<Locatable> expected) throws Exception {
+        final OverlapDetector<Locatable> targetDetector = OverlapDetector.create(input);
+
+        Set<Locatable> actual = targetDetector.getOverlaps(query);
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: final

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        targetDetector.getOverlaps(null);
+    }
+
+    @Test
+    public void testAddTwice() throws Exception {
+        List<Locatable> input = Arrays.asList(
+                new Interval("1",10,100),
+                new Interval("1",10,100)
+        );
+        OverlapDetector.create(input);
```
Comment (yfarjoun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: might as well check that you don't get two overlaps...or do you?

**Comment by yfarjoun** [comment]
LGTM except for one comment and then finals and spaces in the test file. Please have intelliJ fix it up for you. 

all yours @akiezun

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java (hunk 2 above)
Comment (akiezun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: fixed

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java (hunk 3 above)
Comment (akiezun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: IMO this space is useless waste and decreases readability. But fine, i will follow this convention

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java (hunk 4 above)
Comment (akiezun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: done

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java (hunk 5 above)
Comment (akiezun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: done

File: src/main/java/htsjdk/samtools/util/OverlapDetector.java (hunk 6 above)
Comment (akiezun) [inline src/main/java/htsjdk/samtools/util/OverlapDetector.java]: done

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 7 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: done

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 8 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: done

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 9 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: done

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 10 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: another space waster. but fine, done

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 11 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: yup

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 12 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: ok

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 13 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: all 29 are fixed

File: src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java (hunk 14 above)
Comment (akiezun) [inline src/test/java/htsjdk/samtools/util/OverlapDetectorTest.java]: good point. you don't - you get 1 interval back (1:10-100)
